docker/listeners/listeners_unix.go

```diff
@@ -12,12 +12,16 @@ import (
 	"github.com/coreos/go-systemd/activation"
 	"github.com/docker/go-connections/sockets"
 	"github.com/docker/libnetwork/portallocator"
+	"github.com/opencontainers/runc/libcontainer/cgroups/systemd"
 )
 
 // Init creates new listeners for the server.
 func Init(proto, addr, socketGroup string, tlsConfig *tls.Config) (ls []net.Listener, err error) {
 	switch proto {
 	case "fd":
+		if !systemd.UseSystemd() {
+			return nil, fmt.Errorf("can't use 'fd' to create sockets on non-systemd system")
+		}
 		ls, err = listenFD(addr, tlsConfig)
```

Comment:
I think this is not going to work if a system use systemd but not started via systemd, for example, my system `CentOS-7`, it use systemd, but I start the docker daemon via cli `docker daemon -D -H fd://` 
<pre><code>[root@centos-188 lei]# docker daemon -D -H fd://
FATA[0000] No sockets found</code></pre>
`systemd.UseSystemd` is just check if your system use systemd or not, but not 
check if the docker is started by systemd or not.